### PR TITLE
Enhancement on the testcase to cover more possibilities

### PR DESCRIPTION
**Pull request by chendave**
This change enriches the testcase of `TestPriorityQueue_MoveAllToActiveOrBackoffQueue`
to cover the case that pods in the unschedulableQ could be moved to activeQ after the
backing off.

Signed-off-by: Dave Chen <dave.chen@arm.com>

**What type of PR is this?**

Add one of the following kinds:

/kind cleanup

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 1]
```diff
@@ -359,17 +359,38 @@ func TestPriorityQueue_Delete(t *testing.T) {
 }
 
 func TestPriorityQueue_MoveAllToActiveOrBackoffQueue(t *testing.T) {
-	q := NewPriorityQueue(newDefaultQueueSort())
+	c := clock.NewFakeClock(time.Now())
+	q := NewPriorityQueue(newDefaultQueueSort(), WithClock(c))
 	q.Add(&medPriorityPod)
+	q.MoveAllToActiveOrBackoffQueue("test")
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:365]: why is this line necessary?

**Review by alculquicondor — COMMENTED** [review]
Please run the test several times to verify there are no other race conditions:

```
go test -run <testname> -count 100
```

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 1 above)
Comment (soulxu) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:365]: I guess that is for updating the `PriorityQueue.moveRequestCycle`, maybe we can assert  the `PriorityQueue.moveRequestCycle` and `PriorityQueue.schedulingCycle` after this line, that help the people understand the purpose for this line.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 2]
```diff
@@ -359,17 +359,38 @@ func TestPriorityQueue_Delete(t *testing.T) {
...
+	// Make sure the move request is happened first.
+	c.Step(time.Second)
```
Comment (soulxu) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:367]: The next two lines for `AddUnschedulableIfNotPresent`, they depend on the `PriorityQueue.moveRequestCycle` and `PriorityQueue.schedulingCycle` to decide going to the backoff queue or unschedule queue. So this should be useless, it is only useful for `MoveAllToActiveOrBackoffQueue`, it needs the clock to determine the pod is backoff or not.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 3]
```diff
@@ -359,17 +359,38 @@ func TestPriorityQueue_Delete(t *testing.T) {
...
+		t.Errorf("Expected 2 items to be in podBackoffQ, but got: %v", q.podBackoffQ.Len())
 	}
+
+	// pop out the pods in the backoffQ.
+	q.podBackoffQ.Pop()
+	q.podBackoffQ.Pop()
+
+	q.schedulingCycle++
+	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(&unschedulablePod), q.SchedulingCycle())
+	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(&highPriorityPod), q.SchedulingCycle())
```
Comment (soulxu) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:383]: so they are going to unschedulable queue here? this is different with line 368 an 369, so it is better to have assert about it.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 1 above)
Comment (chendave) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:365]: I pulled this line ahead compared with original testcase, combined with the line below `c.Step(time.Second)` just to make sure the move request is happened first, and `p.moveRequestCycle` got updated.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 2 above)
Comment (chendave) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:367]: > So this should be useless, it is only useful for MoveAllToActiveOrBackoffQueue, it needs the clock to determine the pod is backoff or not.

sorry, I am quite follow you here.  In this testcase, all that `MoveAllToActiveOrBackoffQueue` did is set `p.moveRequestCycle`,  I cannot see `MoveAllToActiveOrBackoffQueue` needs the clock anywhere.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 3 above)
Comment (chendave) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:383]: yes.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 2 above)
Comment (soulxu) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:367]: > > So this should be useless, it is only useful for MoveAllToActiveOrBackoffQueue, it needs the clock to determine the pod is backoff or not.
> 
> sorry, I am quite follow you here. In this testcase, all that `MoveAllToActiveOrBackoffQueue` did is set `p.moveRequestCycle`, I cannot see `MoveAllToActiveOrBackoffQueue` needs the clock anywhere.

Sorry, I didn't describe clearly.  :)

I mean `AddUnschedulableIfNotPresent` needn't the clock, so you needn't step the clock here

We only need to step the clock when you test the `MoveAllToActiveOrBackoffQueue `, like line 385. It needs the clock to determine pod going to which queue:

`https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/internal/queue/scheduling_queue.go#L514`

`https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/internal/queue/scheduling_queue.go#L281`

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 3 above)
Comment (chendave) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:382]: Done

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 1 above)
Comment (chendave) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go]: > why is this line necessary?

not necessary actually, have reverted it back.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 2 above)
Comment (chendave) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go]: Done

**Comment by chendave** [comment]
/remove-kind flake

remove the tag since there is low chance to reproduce the flake unless we deliberately add some time consuming operation before the `MoveAllToActiveOrBackoffQueue`, this not true for now.

Big thanks to @alculquicondor  @soulxu  for review!

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 4]
```diff
@@ -359,17 +359,41 @@ func TestPriorityQueue_Delete(t *testing.T) {
...
+		t.Errorf("Expected %v and %v in the unschedulableQ", unschedulablePod.Name, highPriorityPod.Name)
+	}
+	// Move clock by podInitialBackoffDuration, so that pods in the unschedulableQ would pass the backing off,
+	// and the pods will be moved into activeQ.
+	c.Step(q.podInitialBackoffDuration)
+	q.MoveAllToActiveOrBackoffQueue("test")
+	if q.activeQ.Len() != 3 {
+		t.Errorf("Expected 3 items to be in activeQ, but got: %v", q.activeQ.Len())
+	}
+	if q.podBackoffQ.Len() != 0 {
```
Comment (soulxu) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:393]: This assertion seems meanless if without assert the unschedulable queue, since line 378 the pod backoff queue becomes empty, and `AddUnschedulableIfNotPresent` adds the pod to unschedulable queue.  we should assert pods move out unschedulable queue.

**Comment by alculquicondor** [comment]
/lgtm

Did you run the test 100 times?

**Comment by chendave** [comment]
> /lgtm
> 
> Did you run the test 100 times?

yes, run multiple times :)

**Comment by alculquicondor** [comment]
Thanks!

/approve

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 4 above)
Comment (chendave) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:393]: Line @390 should already done that, right? this check is used to compare with Line @372 to tell us the pods are not moved to backoffQ, I think it's not useless.

Comment (soulxu) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:393]: Emm...you are right, sorry about that.
